packages/common/src/codegen/utils/parseSystem.ts

```diff
@@ -10,32 +12,34 @@ export function parseSystem(
   source: string,
   contractName: string,
 ): undefined | { contractType: "contract" | "abstract" } {
-  const ast = parse(source);
-  const contractNode = findContractNode(ast, contractName);
-  if (!contractNode) return;
+  const parser = Parser.create(LanguageFacts.latestVersion());
+  const root = parser.parseFileContents(source).createTreeCursor();
 
-  const contractType = contractNode.kind;
-  // skip libraries and interfaces
-  // we allow abstract systems here so that we can create system libraries from them but without deploying them
-  if (contractType !== "contract" && contractType !== "abstract") return;
+  const contractCursor = findContractOrInterfaceNode(root, contractName);
+  if (!contractCursor) return;
 
-  const isSystem = ((): boolean => {
-    // if using the System suffix, assume its a system
-    if (contractName.endsWith("System") && contractName !== baseSystemName) return true;
+  assertNonterminalNode(contractCursor.node);
+  if (contractCursor.node.kind !== "ContractDefinition") {
+    // it's an interface
+    return;
+  }
+  const contract = new ContractDefinition(contractCursor.node);
+  const contractType = contract.abstractKeyword ? "abstract" : "contract";
 
-    // otherwise check if we're inheriting from the base system
-    let extendsBaseSystem = false;
-    visit(contractNode, {
-      InheritanceSpecifier(node) {
-        if (node.baseName.namePath === baseSystemName) {
-          extendsBaseSystem = true;
-        }
-      },
-    });
-    return extendsBaseSystem && findSymbolImport(ast, baseSystemName)?.path === baseSystemPath;
-  })();
+  // if using the System suffix, assume its a system
+  if (contractName.endsWith("System") && contractName !== baseSystemName) return { contractType };
 
-  if (isSystem) {
-    return { contractType };
+  for (const _ of contractCursor.query([
+    Query.create(`
+      [InheritanceType
+        type_name: [_ ["${baseSystemName}"]]
+      ]
+    `),
+  ])) {
+    if (findSymbolImport(root, baseSystemName)?.path === baseSystemPath) {
```

Comment:
I wonder if this shouldn't be solved using the bindings instead... Worth mentioning, at least.

Reply:
I thought about this as well. There are a couple of issues with this idea:
- MUD processes source files in a file by file basis, which would mean bindings to external symbols cannot be resolved
- for this specific case, bindings would not provide the needed information (ie. the path of the file of the imported symbol)
It may be possible to refactor the project to be able to use Slang's capabilities, but I think that's beyond the scope of an initial migration PR.